discord_markdown_ast_parser/lexer.py

```diff
@@ -1,76 +1,75 @@
 import re
-from dataclasses import dataclass
+from dataclasses import dataclass, field
 from enum import Enum
-from typing import Optional, List, Generator
-
-
-class TokenType(Enum):
-    TEXT_INLINE = 1
-    NEWLINE = 2
-    STAR = 3
-    UNDERSCORE = 4
-    TILDE = 5
-    SPOILER_DELIMITER = 6
-    USER_MENTION = 7
-    ROLE_MENTION = 8
-    CHANNEL_MENTION = 9
-    EMOJI_CUSTOM = 10
-    EMOJI_UNICODE_ENCODED = 11
-    URL_WITH_PREVIEW = 12
-    URL_WITHOUT_PREVIEW = 13
-    QUOTE_LINE_PREFIX = 14
-    CODE_INLINE_DELIMITER = 15
-    CODE_BLOCK_DELIMITER = 16
+from typing import Optional, List, Generator, Dict
+import itertools
 
 
-@dataclass
-class Token:
-    token_type: TokenType
-    value: str
-    groups: Optional[List[str]] = None
+class Lexing:
+    def __init__(self, pattern: Optional[str] = None, flags: re.RegexFlag = re.NOFLAG):
+        self.regex = re.compile(pattern, flags=flags) if pattern else None
+
+    def __call__(self, text: str):
+        return self.regex and self.regex.match(text)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self.regex and self.regex.pattern!r})"
+
+
+URL_REGEX = r"https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)"
```

Comment:
Avoid complicated regexes
```suggestion
URL_REGEX = r"https?:\/\/(www\.)?[-\w@:%.\+~#=]{1,256}\.[a-z]{2,4}\b([-\w@:%.\+~#=?&//]*)"
```